opm/autodiff/ExtendedBlackoilModel_impl.hpp

```diff
@@ -0,0 +1,703 @@
+            V F_solvent = zero_selector.select(ss, ss / (ss + sg)).value();
+
+            const std::vector<int> well_cells(wells().well_cells, wells().well_cells + nperf);
+            const int nw = wells().number_of_wells;
+            V wellSolventFraction = Eigen::Map<const V>(&xw.solventFraction()[0], nperf);
+
+            for (int w = 0; w < nw; ++w) {
+                if(wells().type[w] == PRODUCER) {
+                    for (int perf = wells().well_connpos[w]; perf < wells().well_connpos[w+1]; ++perf) {
+                        wellSolventFraction[perf] = F_solvent[well_cells[perf]];
+                    }
+                }
+            }
+
+            const ADB& rs_perfcells = subset(state.rs, well_cells);
+            int gas_pos = fluid_.phaseUsage().phase_pos[Gas];
+            int oil_pos = fluid_.phaseUsage().phase_pos[Oil];
+            // remove contribution from the dissolved gas.
+            const ADB cq_s_solvent = wellSolventFraction * (cq_s[gas_pos] - rs_perfcells * cq_s[oil_pos]);
+            //cq_s[gas_pos] = ( ones - wellSolventFraction ) *  cq_s[gas_pos];;
+            residual_.material_balance_eq[solvent_pos_] -= superset(cq_s_solvent, well_cells, nc);
+            residual_.material_balance_eq[gas_pos] += superset(cq_s_solvent, well_cells, nc);
+
+        }
+    }
+
+
+
+
+
+
+    template <class Grid>
+    void ExtendedBlackoilModel<Grid>::updateState(const V& dx,
+                                                  ReservoirState& reservoir_state,
+                                                  WellState& well_state)
+    {
+
+        if (has_solvent_) {
+            // Extract solvent change.
+            const int np = fluid_.numPhases();
+            const int nc = Opm::AutoDiffGrid::numCells(grid_);
+            const V zero = V::Zero(nc);
+            const int solvent_start = nc * np;
+            const V dss = subset(dx, Span(nc, 1, solvent_start));
+
+            // Create new dx with the dss part deleted.
+            V modified_dx = V::Zero(dx.size() - nc);
+            modified_dx.head(solvent_start) = dx.head(solvent_start);
+            const int tail_len = dx.size() - solvent_start - nc;
+            modified_dx.tail(tail_len) = dx.tail(tail_len);
+
+            // Call base version.
+            Base::updateState(modified_dx, reservoir_state, well_state);
+
+            // Update solvent.
+            const V ss_old = Eigen::Map<const V>(&reservoir_state.solvent_saturation()[0], nc, 1);
+            const V ss = (ss_old - dss).max(zero);
+            std::copy(&ss[0], &ss[0] + nc, reservoir_state.solvent_saturation().begin());
+
+            // adjust oil saturation
+            const Opm::PhaseUsage& pu = fluid_.phaseUsage();
+            const int pos = pu.phase_pos[ Oil ];
+            for (int c = 0; c < nc; ++c) {
+                reservoir_state.saturation()[c*np + pos] = 1 - ss[c] - reservoir_state.saturation()[c*np + 0] - reservoir_state.saturation()[c*np + 2] ;
```

Comment:
The 0 and 2 indices here, referring to water and gas positions, should be using the phase_pos in the same way as the variable `pos` (should be called `oil_pos` or similar).